File: test/inline.clj
;; `:inline` function metadata -- Clojure's, applied at the call site.
;;
;; The trap this file is built around: an inline that never fires produces
;; exactly the right answer. So every check below is written so that inlining
;; and not inlining give DIFFERENT answers, and the size of the effect is
;; measured separately in `test/inline.mjs` on the deterministic instruction
;; count rather than on a timing.
(require '[clojure.string :as str] '[clojure.edn :as edn])

(defn sh [& args]
  (let [p (.start (ProcessBuilder. (into-array String args)))
        out (slurp (.getInputStream p)) err (slurp (.getErrorStream p))]
    (.waitFor p) {:exit (.exitValue p) :out out :err err}))

(def fails (atom 0))
(defn check [label actual expected]
  (if (= actual expected)
    (println "  ok  " label)
    (do (swap! fails inc)
        (println "  FAIL" label "\n        expected" (pr-str expected)
                 "\n        got     " (pr-str actual)))))
(defn check-that [label ok] (check label (boolean ok) true))

(println ":inline -- the expansion, and what it is allowed to do")

(let [b (sh "./bin/flint" ":src" "test" ":fn" "inline/main" ":out" "out/inline.wasm")]
  (when-not (zero? (:exit b))
    (println "build failed:" (:out b) (:err b)) (System/exit 1)))

(def r (let [x (sh "node" "host/flint.mjs" "out/inline.wasm")]
         (when-not (zero? (:exit x))
           (println "  FAIL the module trapped:" (str/trim (:err x))) (System/exit 1))
         (edn/read-string (str/trim (:out x)))))

;; The inline says something the function does not, so this distinguishes the
;; two paths. Every other check in this file would pass without the feature.
(check "a call site takes the :inline, not the function" (:tattle r) :inlined)
(check "the same var as a VALUE is the function" (:as-value r) :called)
(check "and passed to a higher-order fn it is the function" (:map-value r) [:called :called])

(check ":inline-arities gates it: 2 is inlined" (:gated-2 r) :inline-2)
(check "... and 3 is not, so no argument is silently dropped" (:gated-3 r) :fn-3)

(check "an inlined call still computes what it should" (:add2-2 r) 7)
(check "an un-inlined arity of the same var still works" (:add2-3 r) 12)
(check "syntax quote in an :inline body" (:square r) 49)
(check "an inlined call inside an inlined call" (:nested r) 10)
(check "a local shadowing the var wins over the :inline" (:shadowed r) 11)
(check "both loops in the measurement agree" (:loops-agree r) true)

;; Recorded rather than discovered: the expander receives argument FORMS, so an
;; expansion that mentions one twice evaluates it twice. Clojure has this too.
(check "an :inline that uses its argument twice evaluates it twice"
       (:twice-effect r) [10 2])

(println ":inline -- what the compiler refuses")

(let [x (sh "./bin/flint" ":src" "test/fixtures" ":fn" "inline-loop/main"
            ":out" "out/il.wasm")]
  (check-that "an :inline that re-emits its own name is named, not a stack overflow"
              (and (not (zero? (:exit x)))
                   (str/includes? (str (:out x) (:err x)) "does not terminate")))
  (check-that "... and the message says which var"
              (str/includes? (str (:out x) (:err x)) "inline-loop/spin")))

(let [x (sh "./bin/flint" ":src" "test/fixtures" ":fn" "inline-bad/main"
            ":out" "out/ib.wasm")]
  (check-that "an :inline that is not a function is refused at compile time"
              (and (not (zero? (:exit x)))
                   (str/includes? (str (:out x) (:err x)) "is not a function"))))

(if (pos? @fails)
  (do (println "inline:" @fails "failed") (System/exit 1))
  (println "inline: ok"))

File: test/inline.mjs
// `:inline` measured, not assumed.
//
// The correctness half is in `test/inline.clj`. This is the half that says the
// feature does anything: two loops computing the same answer, one through vars
// with `:inline` and one through identical vars without, compared on the
// DETERMINISTIC instruction count (0009). Equal counts mean the inline did not
// fire, and that is a failure -- an inline that silently does nothing gives the
// right answer every time.
import { load, instantiate } from '../host/flint.mjs';

let fails = 0;
const ok = (label, cond, detail) => {
  if (cond) console.log('  ok   ' + label);
  else { fails++; console.log('  FAIL ' + label + (detail ? '\n        ' + detail : '')); }
};

const { module } = await load(process.argv[2] ?? 'out/inline.wasm');
const run = (mode, iters) => {
  const inst = instantiate(module);
  inst.exports.set_step_limit(0, 1e9);
  const r = inst.main(mode, String(iters));
  if (r.code !== 0) throw new Error(`${mode} failed: ${r.out}`);
  return { out: r.out, steps: Number(inst.exports.stat_steps()) };
};

console.log(':inline -- the instruction count, which is deterministic');

// Per-iteration cost: the same mode at N and at 0, differenced. Comparing two
// modes directly also measures this file's own dispatch, which is a handful of
// instructions and would be reported as an effect of inlining.
const ITERS = 2000;
const loopCost = (mode) => {
  const hot = run(mode, ITERS), cold = run(mode, 0);
  return { steps: hot.steps - cold.steps, out: hot.out };
};
const a = loopCost('inlined'), b = loopCost('plain');
ok('the two loops agree on the answer', a.out === b.out, `${a.out} vs ${b.out}`);
ok('the inlined loop is a real run, not an empty one', a.steps > 10000, String(a.steps));
ok('inlining removed instructions', a.steps < b.steps,
   `inlined ${a.steps}, plain ${b.steps} -- equal counts mean the inline never fired`);

const saved = b.steps - a.steps;
console.log(`    inlined ${a.steps.toLocaleString('en-US')}, ` +
            `plain ${b.steps.toLocaleString('en-US')}, ` +
            `${(100 * saved / b.steps).toFixed(1)}% fewer instructions ` +
            `(${(saved / ITERS).toFixed(1)} per iteration over ${ITERS} iterations, ` +
            `for 2 inlined calls per iteration)`);

// The count is deterministic, so a second run must produce the same number. If
// it does not, the measurement above is not a measurement.
ok('the count repeats exactly', loopCost('inlined').steps === a.steps, 'it did not');

process.exit(fails ? 1 : 0);
